Mathlib/Tactic/NormNum/RealSqrt.lean

```diff
@@ -0,0 +1,123 @@
+/-
+Copyright (c) 2025 Frédéric Dupuis. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Frédéric Dupuis
+-/
+import Mathlib.Data.Real.Sqrt
+
+/-! # `norm_num` extension for `Real.sqrt`
+
+This module defines a `norm_num` extension for `Real.sqrt` and `NNReal.sqrt`.
+
+## Implementation notes
+
+While the extension for `Real.sqrt` can handle both integers and rationals, the one for
+`NNReal.sqrt` can only deal with integers, due to a limitation of norm_num
+(i.e. the `IsRat` type requires a Ring instance).
+-/
+
+namespace Tactic.NormNum
+
+open Qq Lean Lean.Meta Elab.Tactic Mathlib.Meta.NormNum NNReal
+
+lemma isNat_realSqrt {x : ℝ} {nx ny : ℕ} (h : IsNat x nx) (hy : ny * ny = nx) :
+    IsNat √x ny := ⟨by simp [h.out, ← hy]⟩
+
+lemma isNat_nnrealSqrt {x : ℝ≥0} {nx ny : ℕ} (h : IsNat x nx) (hy : ny * ny = nx) :
+    IsNat (NNReal.sqrt x) ny := ⟨by simp [h.out, ← hy]⟩
+
+lemma isNat_realSqrt_neg {x : ℝ} {nx : ℕ} (h : IsInt x (Int.negOfNat nx)) :
+    IsNat √x 0 := by
+  refine ⟨?_⟩
+  simp only [Nat.cast_zero]
+  refine Real.sqrt_eq_zero_of_nonpos ?_
+  simp [h.out]
+
+lemma isNat_realSqrt_neg_of_isRat {x : ℝ} {num : ℤ} {denom : ℕ} (hnum : num ≤ 0)
+    (h : IsRat x num denom) : IsNat √x 0 := by
+  refine ⟨?_⟩
+  simp only [Nat.cast_zero]
+  refine Real.sqrt_eq_zero_of_nonpos ?_
+  let .mk inv eq := h
+  simp only [eq]
+  refine mul_nonpos_of_nonpos_of_nonneg (by exact_mod_cast hnum) ?_
+  rw [invOf_nonneg]
+  exact Nat.cast_nonneg' denom
+
+lemma isRat_realSqrt {x : ℝ} {n sn : ℤ} {d sd : ℕ} (hn : sn * sn = n)
+    (hn₂ : 0 ≤ sn) (hd : sd * sd = d) (hd₂ : sd ≠ 0) (h : IsRat x n d) :
+    IsRat √x sn sd := by
+  refine .mk (invertibleOfNonzero (by exact_mod_cast hd₂)) ?out
+  let .mk inv eq := h
+  rw [eq, ← hn, invOf_eq_inv, ← hd]
+  simp only [Int.cast_mul, Nat.cast_mul, mul_inv_rev, invOf_eq_inv]
+  rw [Real.sqrt_mul (mul_self_nonneg ↑sn)]
+  aesop
+
+/-- `norm_num` extension that evaluates the function `Real.sqrt`. -/
+@[norm_num √_]
+def evalRealSqrt : NormNumExt where eval {u α} e := do
+  let .app _ (x : Q(ℝ)) ← whnfR e | failure
+  have : u =QL 0 := ⟨⟩; have : $α =Q ℝ := ⟨⟩; have : $e =Q √$x := ⟨⟩
```

Comment:
@dupuisf, a trick for next time: if you match without Qq, you can give back all the hints you get from proper matching with `=QL` and `=Q`.

Reply:
Looks like the unused haves linter doesn't like this trick though!